### PR TITLE
Bucm/collection perms

### PR DESCRIPTION
**Pull request by tsmacdonald**
Adds a new `/moderate/` collection permission, that works analogously to `/read/`. Moderate gives read, write and the (new) ability to create moderation reviews for items in a collection.

### CODE REVIEW

**Comment by tsmacdonald** [comment]
Tagging @jeff-bruemmer for the doc changes (renaming the "Curate" collection permission to "Edit"). I also noticed some inconsistency re. capitalization and "permission" vs. "permissions".

**Review by jeff-bruemmer — APPROVED** [review]
Docs look good to me. Thanks for minding the documentation, @tsmacdonald. I've made a note to update the Learn article on collection permissions to reflect this change once bucm lands.

File: src/metabase/models/permissions.clj
Comment (tsmacdonald) [inline src/metabase/models/permissions.clj:269]: This is pretty futzy code and probably the most important function in the entire PR. I'm sure a student of predicate logic and string hacks could tighten it up a bit, but I thought being explicit about some of the different cases was helpful

File: src/metabase/db/data_migrations.clj
Comment (dpsutton) [inline src/metabase/db/data_migrations.clj:394]: these `defmigration`s are always not great if there's any other way. Could you use a normal migration and update the admin group with this Permission and ensure that when created initially with `(perm-group/admin)` (which can create it if it doesn't exist) it adds the Permission here? That way you either add it if the perm group exists, or don't add it in the migration but creating the admin permission group ensures this permission exists. As far as i can tell the id should be where the PermissionGroup name is `"Administrators"`

File: docs/administration-guide/06-collections.md
Comment (camsaul) [inline docs/administration-guide/06-collections.md:31]: Shouldn't there be a new entry here about "Moderate" permissions?

File: src/metabase/db/data_migrations.clj
Comment (camsaul) [reply src/metabase/db/data_migrations.clj:394]: The permissions system is prefix-based in the sense that if you have permissions to `/some/thing/` then you also have perms to anything starting with `/some/thing/`, e.g. `/some/thing/else/`. Admins all have permissions to `/`, which implies `/collection/root/moderate`... this line shouldn't be needed.

Seconding what Dan said about adding `defmigrations` as well -- we really really really shouldn't add them unless we absolutely can't do it any other way -- you could pretty easily do this as single SQL query if you needed to and put it in Liquibase

File: src/metabase/models/permissions.clj
Comment (camsaul) [reply src/metabase/models/permissions.clj:269]: Definitely not a big fan of doing special-case hacks to the permissions system. The whole house of cards is going to come crashing down if we invert how things work, since other things rely on the prefix-based nature of the perms system (such as DB queries that do `LIKE /something/%`. **Please** don't add this stuff 😬

Comment (camsaul) [reply src/metabase/models/permissions.clj:269]: I don't think the whole concept of "Moderate permission has access to everything in that path" makes any sense.

There are two ways you can modify a Collection: 

-  "Editing" it by adding or removing items, changing the name, etc.
- Add reviews

Both of these are subsets of "write" perms for a Collection. So they should both go under a common prefix, and then you can give people one or the other or just the entire prefix if you want them to have both. For example you could do something like

```
/collection/1/        -> full perms to do anything for Collection 1
/collection/1/edit/   -> perms to edit collection and add or remove items to Collection 1
/collection/1/review/ -> perms to add/remove reviews for Collection 1
```

Now assuming you want the permission level to be are none < read-only < "Edit" < "Moderate" you should grant permissions as follows:

```
Read perms     -> /collection/1/read (no perms to /edit/ or /review/
Edit perms     -> /collection/1/edit/ and /collection/1/read/ (perms to read, and to /edit/)
Moderate perms -> /collection/1/ (full perms to do anything to Collection 1, including read, edit, and review)
```

Then you can go around and check whether people have `/collection/1/review/` perms, which is true whether they're an admin (they have `/` perms) or someone else you granted moderate perms to (they have `/collection/:id/` perms).

You'll also want to write a migration to migrate existing `/collection/:id/` perms in the DB to `/collection/:id/edit/` since the meaning has changed. This should be easy to do tho. You should definitely do it as a Liquibase migration instead of a Clojure-land one because the Clojure-land stuff is so fragile

Comment (camsaul) [reply src/metabase/models/permissions.clj:269]: Thinking about this a bit more.

You might consider tweaking the paths a bit so you don't need to grant two entries to give someone Read + "Edit" perms. Read perms seem like a subset of whatever "Edit" perms are.

That was previously reflected by having `/collection/1/` be the common prefix -- `/collection/1/` meant  "readwrite" and `/collection/1/read/` meant "read", a subset of "readwrite". Maybe the new paths should look like

```
/collection/1/readwrite/read/ -> perms to Read 
/collection/1/readwrite/      -> perms to Edit, and to Read
/collection/1/review/         -> perms to add/remove reviews 
/collection/1/                -> perms to do anything, including Edit, Read, or "Moderate"
```

I'm cool with whatever approach as long as we don't flip the whole prefix concept on its head

**Review by camsaul — CHANGES_REQUESTED** [review]
I don't think you need to fundamentally rework the way the permissions system itself works in order to get this working. I really don't want to toss the basic prefix-based stuff that we've used in the past out the window. In the long run these sorts of special cases are going to be almost impossible to reason about once they start accumulating, and it's going to lead to nasty bugs. Perms bugs are some of the worst kinds of bugs, since we have to do the whole security patch routine.

I suggested a few ways to implement the new perms that don't require changing the permissions system itself... try that out and let me know if you can't get it working.

Also: please no new data migrations unless there is absolutely no other way to do what you want

File: src/metabase/models/permissions.clj
Comment (camsaul) [reply src/metabase/models/permissions.clj:269]: Another alternative to consider:

for a few models we consider you to have "read" perms if you have "partial perms" for an object, e.g. any perms starting with that object's prefix. e.g.

```clj
(can-read? Collection 1) 
-> 
(user-has-partial-perms? "/collection/1/") 
-> 
user has any perms starting with /collection/1/
```

You could set it up this way and then stick with `/collection/1/read/`, `/collection/1/edit/` and `/collection/1/review/` and not have to worry about migrating `/read/` to `/readwrite/read`, or granting read + edit perms at the same time.

I'm leaning towards this approach out of the 3 I've suggested since it would require the least amount of changes. I'm cool with any approach that keeps the spirit of the perms system intact tho.

Comment (tsmacdonald) [reply src/metabase/models/permissions.clj:270]: Thanks for the thoughtful feedback. To synthesize, it sounds like the easiest fix is:

* Keep the new `/collection/:id/moderate/` as-is. Moderation necessarily implies readwrite access.
* Keep the old `/collection/:id/read/` as-is
* Introduce `/collection/:id/edit/` to mark readwrite (but not moderation) permission.
* `/collection/:id/` will be the "do-everything" permission, but in practice it just means that admins who have `/` will be able to do what they want.

However, this doesn't solve the problem of needing permissions to be a strict prefix of what they grant access too. E.g., a permission of `LIKE '/collection/42/moderate/%'` won't work (but should) if the item path is `/collection/42/edit`.

As far as I can figure, the only way to make prefixing work is to stack it. So the paths would be:

| access level | path |
|--|--|
| read collection 42 | `/collection/42/moderate/edit/read/`
| edit collection 42 (implies read access) | `/collection/42/moderate/edit/`
| moderate collection 42 (implies read+edit access) | `/collection/42/moderate/`

This is intuitively a little gross to me, and more specifically I think it's brittle (it'll be a PITA if we add a new type of permission) and I would prefer not to write a DB migration to change every single path in the DB (i.e., `collection/42/read/` → `collection/42/moderate/edit/read/`)

If there's any way to drop the prefix requirement, I think everything would be cleaner. I would have thought that permission-checking should happen at the application layer (or at least with carefully-tuned functions that spit out the right SQL) and hope we don't have ad hoc SQL being written all over the place. IMHO good encapsulation is to only deal with the exact contents of the `permission` table in `models/permission.clj` and let the rest of the codebase deal with the `can-read?` / `can-edit?` / `can-moderate?` functions.

Comment (camsaul) [reply src/metabase/models/permissions.clj:270]: I think our wires are a bit crossed. Let me try to explain what I was proposing a little more clearly.

For perms that you'd _grant_, you'd grant:

| access level | path |
|--|--|
| Read-only access to Collection 1 | `/collection/1/read/` |
| "Edit" Collection 1 | `/collection/1/edit/` |
| "Full" perms for Collection 1 including edit or 'moderate | `/collection/1/` |

For perms that you'd _check_, you'd check

| Perms to Check | How to Check | (SQL equivalent |
|--|--|--|
| Current user can read? | current user has partial (any) perms? | `perms LIKE '/collection/1/%'` |
| Current user can edit? | current user has edit perms? | `concat(perms, '%') LIKE '/collection/1/edit/'` |
| Current user can 'review (moderate?)' | current user has 'review' (?) perms | `concat(perms, '%') LIKE '/collection/1/review/'` |

So there's no special cases where presence of any particular permission implies any other. We already have tooling for the "partial perms" stuff (and use it in several other places) as well as code for generating the sorts of SQL in the example above. So you could do all of this without changing any existing tooling. The only thing you'd have to change is migrating existing perms like `/collection/1/` to `/collection/1/edit/`, but you can do that with a simple SQL statement in Liquibase

Comment (tsmacdonald) [reply src/metabase/models/permissions.clj:270]: Makes sense, many thanks!

File: docs/administration-guide/06-collections.md
Comment (tsmacdonald) [reply docs/administration-guide/06-collections.md:31]: @jeff-bruemmer is planning to write it up for the 41 release

File: src/metabase/db/data_migrations.clj
Comment (tsmacdonald) [reply src/metabase/db/data_migrations.clj:394]: :sparkles: removed, `/` permission will be adequate

File: src/metabase/models/collection/graph.clj
Comment (tsmacdonald) [inline src/metabase/models/collection/graph.clj:45]: change this, Tim

**Review by camsaul — APPROVED** [review]
I didn't go over every line in crazy detail but I looked over this version for 15 minutes or so and it looks sensible to me. I was hoping we'd get some more eyes from someone else on the backend team on it as well, but I don't think we should hold up on merging it any longer

**Pull request merged**
